google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/facet/flexible/AppEngineFlexibleSupportProvider.java

```diff
@@ -0,0 +1,170 @@
+import com.intellij.remoteServer.impl.configuration.deployment.DeployToServerRunConfiguration;
+import com.intellij.util.containers.ContainerUtil;
+
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import javax.swing.JComponent;
+import javax.swing.JPanel;
+
+/**
+ * Adds Flexible support to new or existing IJ modules.
+ */
+public class AppEngineFlexibleSupportProvider extends FrameworkSupportInModuleProvider {
+
+  @NotNull
+  @Override
+  public FrameworkTypeEx getFrameworkType() {
+    return AppEngineFlexibleFrameworkType.getFrameworkType();
+  }
+
+  @NotNull
+  @Override
+  public FrameworkSupportInModuleConfigurable createConfigurable(
+      @NotNull FrameworkSupportModel model) {
+    return new AppEngineFlexibleSupportConfigurable();
+  }
+
+  @Override
+  public boolean isEnabledForModuleType(@NotNull ModuleType moduleType) {
+    return moduleType instanceof JavaModuleType;
+  }
+
+  @Override
+  public boolean isSupportAlreadyAdded(@NotNull Module module,
+      @NotNull FacetsProvider facetsProvider) {
+    return !facetsProvider.getFacetsByType(module, AppEngineFlexibleFacetType.ID).isEmpty()
+        || !facetsProvider.getFacetsByType(module, AppEngineStandardFacet.ID).isEmpty();
+  }
+
+  static class AppEngineFlexibleSupportConfigurable extends FrameworkSupportInModuleConfigurable {
+
+    private JPanel mainPanel;
+    private CloudSdkPanel cloudSdkPanel;
+
+    @Nullable
+    @Override
+    public JComponent createComponent() {
+      return mainPanel;
+    }
+
+    @Override
+    public void addSupport(@NotNull Module module, @NotNull ModifiableRootModel rootModel,
+        @NotNull ModifiableModelsProvider modifiableModelsProvider) {
+      FacetType<AppEngineFlexibleFacet, AppEngineFlexibleFacetConfiguration> facetType =
+          AppEngineFlexibleFacet.getFacetType();
+      AppEngineFlexibleFacet facet = FacetManager.getInstance(module).addFacet(
+          facetType, facetType.getPresentableName(), null /* underlyingFacet */);
+
+      // Allows suggesting app.yaml and Dockerfile locations in facet and deployment UIs.
+      VirtualFile[] contentRoots = rootModel.getContentRoots();
+      AppEngineProjectService appEngineProjectService = AppEngineProjectService.getInstance();
+      if (contentRoots.length > 0) {
+        facet.getConfiguration().setYamlPath(
+            appEngineProjectService.getDefaultAppYamlPath(contentRoots[0].getPath()));
+        facet.getConfiguration().setDockerfilePath(
+            appEngineProjectService.getDefaultDockerfilePath(contentRoots[0].getPath()));
+      }
+
+      // Only adds deployment run configuration for now. Stackdriver debugger to follow.
+      setupDeploymentRunConfiguration(module, facet);
+
+      CloudSdkService sdkService = CloudSdkService.getInstance();
+      if (!sdkService.validateCloudSdk(cloudSdkPanel.getCloudSdkDirectoryText())
```

Comment:
Not if we don't explicitly call apply(). I thought it'd be cleaner this way, without having to deal with catching ConfigurationException.